File: generator-common/src/main/java/io/github/potjerodekool/openapi/common/generate/ValidationExtensions.java
package io.github.potjerodekool.openapi.common.generate;

import io.github.potjerodekool.openapi.common.generate.model.Digits;

import java.util.Map;
import java.util.Optional;

public final class ValidationExtensions {

    public static final String ASSERT = "x-assert";

    private ValidationExtensions() {
    }

    public static Optional<Digits> digits(final Map<String, Object> extensions) {
        if (extensions == null) {
            return Optional.empty();
        }

        final var validation = getValidation(extensions);
        final var digits = validation.get("digits");

        if (!(digits instanceof Map)) {
            return Optional.empty();
        }

        final var digitsMap = (Map<String, Object>) digits;

        final var integer = (Integer) digitsMap.get("integer");
        final var fraction = (Integer) digitsMap.get("fraction");
        return Optional.of(new Digits(integer, fraction));
    }

    public static Object allowedValue(final Map<String, Object> extensions) {
        final var validation = getValidation(extensions);
        return validation.get("allowed-value");
    }

    public static Map<String, Object> getValidation(final Map<String, Object> extensions) {
        return extensions != null
                ? (Map<String, Object>) extensions.getOrDefault("x-validation", Map.of())
                : Map.of();
    }
}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/generate/ValidationExtensions.kt
package io.github.potjerodekool.openapi.common.generate

import io.github.potjerodekool.openapi.common.generate.model.Digits
import java.util.*

object ValidationExtensions {
    const val ASSERT: String = "x-assert"

    @JvmStatic
    fun digits(extensions: MutableMap<String?, Any>?): Optional<Digits> {
        if (extensions == null) {
            return Optional.empty()
        }

        val validation = getValidation(extensions)
        val digits = validation["digits"] as? MutableMap<*, *>
            ?: return Optional.empty()

        val digitsMap = digits as MutableMap<String, Any>

        val integer = digitsMap["integer"] as Int?
        val fraction = digitsMap["fraction"] as Int?
        return Optional.of(
            Digits(
                integer!!, fraction!!
            )
        )
    }

    @JvmStatic
    fun allowedValue(extensions: MutableMap<String?, Any>?): Any? {
        val validation = getValidation(extensions)
        return validation["allowed-value"]
    }

    @JvmStatic
    fun getValidation(extensions: MutableMap<String?, Any>?): Map<String, Any> {
        return if (extensions != null
        ) extensions.getOrDefault("x-validation", java.util.Map.of<Any, Any>()) as Map<String, Any>
        else java.util.Map.of()
    }
}
